src/qibo/hamiltonians/hamiltonians.py

```diff
@@ -51,60 +76,50 @@ def from_symbolic(cls, symbolic_hamiltonian, symbol_map):
 
     def eigenvalues(self, k=6):
         if self._eigenvalues is None:
-            self._eigenvalues = self.K.eigvalsh(self.matrix, k)
+            self._eigenvalues = self.backend.calculate_eigenvalues(self.matrix, k)
         return self._eigenvalues
 
     def eigenvectors(self, k=6):
         if self._eigenvectors is None:
-            self._eigenvalues, self._eigenvectors = self.K.eigh(self.matrix, k)
+            self._eigenvalues, self._eigenvectors = self.backend.calculate_eigenvectors(self.matrix, k)
         return self._eigenvectors
 
     def exp(self, a):
         if self._exp.get("a") != a:
             self._exp["a"] = a
-            if self._eigenvectors is None or self.K.issparse(self.matrix):
-                self._exp["result"] = self.K.expm(-1j * a * self.matrix)
-            else:
-                expd = self.K.diag(self.K.exp(-1j * a * self._eigenvalues))
-                ud = self.K.transpose(self.K.conj(self._eigenvectors))
-                self._exp["result"] = self.K.matmul(
-                    self._eigenvectors, self.K.matmul(expd, ud))
+            self._exp["result"] = self.backend.calculate_exp(a, self._eigenvectors,
+                                                             self._eigenvalues, self.matrix)
         return self._exp.get("result")
 
     def expectation(self, state, normalize=False):
-        if isinstance(state, states.AbstractState):
-            ev = state.expectation(self, normalize)
-        elif isinstance(state, K.tensor_types):
+        if isinstance(state, self.backend.tensor_types):
             shape = tuple(state.shape)
             if len(shape) == 1: # state vector
-                from qibo.core.states import VectorState
-                sobj = VectorState.from_tensor(state)
+                return self.backend.calculate_expectation_state(self.matrix, state, normalize)
             elif len(shape) == 2: # density matrix
-                from qibo.core.states import MatrixState
-                sobj = MatrixState.from_tensor(state)
+                return self.backend.calculate_expectation_density_matrix(self.matrix, state, normalize)
             else:
                 raise_error(ValueError, "Cannot calculate Hamiltonian "
                                         "expectation value for state of shape "
                                         "{}.".format(shape))
-            ev = sobj.expectation(self, normalize)
         else:
             raise_error(TypeError, "Cannot calculate Hamiltonian expectation "
                                    "value for state of type {}."
                                    "".format(type(state)))
-        return ev
 
     def eye(self, n=None):
         if n is None:
             n = int(self.matrix.shape[0])
-        return self.K.eye(n, dtype=self.matrix.dtype)
+            #TODO: see if we can remove backend.eye method
+        return self.backend.eye(n, dtype=self.matrix.dtype)
```

Comment:
I've noticed that this is not fully correct because the argument for the identity matrix is the number of qubits:
https://github.com/qiboteam/qibo/blob/fad9755003c2b018953576cc5365f5b0b237efb9/src/qibo/backends/matrices.py#L62-L63
Theoretically I could replace the argument n with log2(n) and everything should work. I don't know if at this point it is reasonable to reintroduce the eye method just to have these different options:
* matrices.I(n): creates a 2^n x 2^n identity matrix
* backend.eye(n): creates a n x n identity matrix

Reply:
> Another solution is just to change the argument in the identity matrix since it could be misleading. Let me know what you think @stavros11.

Yes, I agree. We should just change `matrices.I(n)` so that `n` is the dimension of the matrix and not the number of qubits, which is also more flexible. You can do this here, or I can do it in the base branch. Let me know what you prefer.

Reply:
I fixed this in the base branch and I am propagating it to all open branches. If you merge the latest version matrices.I(n) should create a n x n identity.